torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -207,115 +208,19 @@ class ProcessGroupNCCL : public ProcessGroup {
   // enables synchronizing the appropriate streams and avoids stalling PyTorch's
   // default stream while running the callback. In case of multiple then
   // callbacks, each will be executed on its own fresh stream.
-  struct FutureNCCL : at::ivalue::Future {
+  struct FutureNCCL : at::cuda::CUDAFuture {
    public:
-    explicit FutureNCCL(
+    FutureNCCL(
         at::IValue value,
         std::shared_ptr<std::vector<at::cuda::CUDAEvent>> cudaEvents)
-        : at::ivalue::Future(c10::ListType::create(c10::TensorType::get())),
-          cudaEvents_(std::move(cudaEvents)) {
+        : at::cuda::CUDAFuture(c10::ListType::create(c10::TensorType::get())){
+      cudaEvents_ = std::move(cudaEvents);
```

Comment:
Are these events used anywhere? Will the `markCompleted` call below this line immediately override these `cudaEvents_`?

Reply:
These events will not be overridden when the future is marked complete, because FutureNCCL provides its own `postMarkCompletedHook` which does nothing (and which overrides the one from CUDAFuture, which is where we store the events).

However, these events will still be used by the postWaitHook, by the wrapCallback, etc.

It's somewhat hacky, but it works, and it's a temporary solution only until we finish merging Future and Work (at which point ProcessGroupNCCL should be able to use CUDAFuture unmodified).